Add concrete-core fixtures

aPere3 opened the pull request:
### Resolves
zama-ai/concrete_internal#265

### Description
This PR modifies the previously existing `concrete-test` application to create two new libraries:
+ A `concrete-core-fixture` library which contains the necessary tools to create fixtures for the `*Engine` traits of `concrete-core`
+ A `concrete-core-test` library which uses the `concrete-core-fixture` to test different backends

As examples, fixtures for the `CleartextCreationEngine` and `LweCiphertextEncryptionEngine` were added.

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (apply the `aws_test` to the PR to launch the tests on AWS)
* [x] The draft release description has been updated
* [x] Check for breaking changes and add them to commit message following the conventional commit [specification][conventional-breaking]

<!--
### Requires: `<link_your_required_issue_here>`
-->

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

tmontaigu commented on concrete-core-fixture/src/fixture/mod.rs:
```diff
@@ -0,0 +1,179 @@
+
+    /// A type containing prototypical values (such as secret keys) passed around the engine
+    /// execution.
+    type Bypass;
```
calling it `Bypass` feels a bit strange, this type seems to be forwarded to `prepare_context` so I'd suggest `ForwardedContext` but I'm really not set.

aPere3 replied:
Actually the `Bypass` type contains prototypes of secret keys which needs to be forwarded to the `process_context` method to decrypt the output of the engine execution, but will _bypass_ the `execute_engine` method.

I would prefer not to use the `Context` word for this one, because it is not part of the execution context of the _engine_ (not forwarded to `execute_engine`). Something in the spirit of `Keys` or `SecretKeys` or `KeysProto` or `SecretKeysPrototypes` may be better suited. 

Waiting for your inputs on this one :wink:

agnesLeroy replied:
@aPere3 they're not key prototypes are they? They're `SynthesizedKeys` rather? Or do I get it wrong?

aPere3 replied:
No no, they are prototypes. Btw, if the `Prototypes` and `Synthesizing`  words do not work for you, I am completely open to something else .

agnesLeroy replied:
After today's discussion and your presentation I understand better :smile: Ok so `SecretKeyPrototypes` sounds good to me.

aPere3 replied:
Ok done !

agnesLeroy commented on concrete-core-fixture/src/fixture/mod.rs:
```diff
@@ -0,0 +1,179 @@
+//! down to:
+//!
+//! + Iterating over a set of parameters multiple times
+//! + Generating random raw inputs (where _raw_ ~ unsigned integers derivatives)
```
Maybe use `designates` instead of a symbol?

aPere3 replied:
I modified that to be a bit more explicit:
> where _raw_ is a type derived of rust´s integers like u32, u64, Vec<u32>, etc

aPere3 commented on concrete-core-fixture/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "concrete-core-fixture"
```
Oh yes, you are right, I totally forgot about all that. Will do !

aPere3 commented on concrete-core-fixture/src/fixture/mod.rs:
```diff
@@ -0,0 +1,179 @@
+        repetitions: Repetitions,
+        sample_size: SampleSize,
+    ) -> bool {
+        assert!(repetitions.0 > 0);
+        (0..repetitions.0)
+            .map(|_| {
+                Self::test(
+                    maker,
+                    engine,
+                    parameters,
+                    &Self::generate_random_raw_inputs(parameters),
+                    sample_size,
+                )
+            })
+            .reduce(BitAnd::bitand)
+            .unwrap()
+    }
```
Smart :wink: Will do !

aPere3 replied:
Its actually better because we should have done `assert(repetitions > 1)` to ensure that the reduce does not panic.

aPere3 commented on concrete-core-fixture/src/generation/prototyping/lwe_bootstrap_key.rs:
```diff
@@ -0,0 +1,89 @@
+impl PrototypesLweBootstrapKey<Precision32, BinaryKeyDistribution, BinaryKeyDistribution>
+    for Maker
+{
+    type LweBootstrapKeyProto = ProtoBinaryBinaryLweBootstrapKey32;
```
We can !

aPere3 commented on concrete-core-fixture/src/generation/prototyping/lwe_ciphertext_vector.rs:
```diff
@@ -0,0 +1,95 @@
+        Precision = Precision,
+        KeyDistribution = KeyDistribution,
+    >;
+    fn encrypt_plaintext_vector_to_lwe_ciphertext_vector(
```
Probably yes !

aPere3 replied:
Same goes for glwe ciphertexts actually. Thanks for noticing ;)

aPere3 commented on concrete-core-test/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "concrete-core-test"
```
Done, thanks !

aPere3 commented on concrete-core-test/src/core.rs:
```diff
@@ -0,0 +1,42 @@
+        paste!{
+            #[test]
+            fn [< test_ $fixture:snake _ $precision:snake _ $($types:snake)_+ >]() {
+                let mut maker = Maker::default();
```
Yes exactly. We need the core engine in the maker to manipulate the prototypical entities, and synthesize the concrete ones. When we will add datatypes which resides in separate hardware memories, we will also need to put the engine allowing to transfer them in the Maker. 

Does that seem reasonable to you ?

blowfish880 commented on concrete-core-fixture/src/generation/mod.rs:
```diff
@@ -0,0 +1,64 @@
+//! Generating the pre-execution context for the fixtures can be tricky. We have to generate
+//! entities of the right type, which are also functionally compatible (encoded with the same secret
+//! key for instance). The [`Maker`] type is the main entry point to context setup, which is a two
+//! steps process:
```
a two-step process

aPere3 replied:
Done thanks !

aPere3 commented on concrete-core/src/backends/core/private/crypto/lwe/tests.rs:
```diff
@@ -20,6 +23,89 @@ use crate::backends::core::private::test_tools::{
     random_uint_between,
 };
 
+fn test_keyswitch<T: UnsignedTorus + RandomGenerable<UniformMsb>>() {
```
No idea what happenned, I'll remove it.

aPere3 commented on concrete-core/src/specification/entities/markers.rs:
```diff
@@ -73,7 +73,7 @@ entity_kind_marker! {
 /// [`KeyDistributionMarker`] types are only defined in the specification part of the library, and
 /// can not be defined by a backend.
-pub trait KeyDistributionMarker: seal::KeyDistributionMarkerSealed {}
+pub trait KeyDistributionMarker: seal::KeyDistributionMarkerSealed + 'static {}
```
Basically what this means is that a type which implements `KeyDistributionMarker` will own its data, i.e. will not contain references to things which are non-static. We need it for a patch which uses the current npe (which does not use the same types for key distributions, hence we have to do a conversion). But it is not in this PR actually, it is for the test of the keyswitch. I think I should remove it from here and move it to the commit which add the keyswitch test.

agnesLeroy commented:
@aPere3 I'm thinking that we should be modifying the CI together with this PR to be able to launch the tests from Github.

aPere3 commented:
@agnesLeroy, yes, I did just that ! Normally everything should be good now !

blowfish880 commented on concrete-core-fixture/src/fixture/mod.rs:
```diff
@@ -0,0 +1,177 @@
+use concrete_core::prelude::AbstractEngine;
+use std::ops::BitAnd;
+
+/// A trait for types implementing a fixture for a particular engine traits.
```
for a particular engine trait.

blowfish880 commented on concrete-core-fixture/src/raw/mod.rs:
```diff
@@ -0,0 +1,8 @@
+//! A module manipulating raw messages.
+//!
+//! For all the fixtures, we need to be able to generate input plaintext, and analyze output
```
input plaintexts

blowfish880 commented on concrete-core-fixture/src/raw/mod.rs:
```diff
@@ -0,0 +1,8 @@
+//! A module manipulating raw messages.
+//!
+//! For all the fixtures, we need to be able to generate input plaintext, and analyze output
+//! plaintexts. We implement those generation and analysis function only for the _raw_ `u32` and
```
generation and analysis functions

blowfish880 commented on concrete-core-fixture/src/generation/synthesizing/mod.rs:
```diff
@@ -0,0 +1,30 @@
+//! A module to synthesize actual entities from prototypical entities.
+//!
+//! This module allows to convert back and forth between prototypical entities to the actual entity
```
back and forth between prototypical entities and the actual entity types

agnesLeroy commented:
@aPere3 just so we remember: it would be good to have README files in the two crates, and to have the main commands to launch the tests in the test crate.

blowfish880 reviewed (commented):
I think I understand the high level structure of the new testing framework and it makes sense to me at this point. Apart from a few typos I didn't find anything to complain about.

aPere3 commented:
Thanks @blowfish880 for your review ! I corrected your points, everything should be good now. @agnesLeroy, I added READMEs in both folders. I rebased on master, which fixes the new clippy lint, so everything should work now.
